include/oneapi/dpl/pstl/hetero/utils_hetero.h

```diff
@@ -0,0 +1,152 @@
+struct adjacent_find_fn
+{
+    _Predicate __predicate;
+
+    // the functor is being used instead of a lambda because
+    // at this level we don't know what type we get during zip_iterator unpack
+    // whereas lambdas with auto in arg parameters are supported since C++14
+    template <typename _Pack>
+    bool
+    operator()(const _Pack& __packed_neighbor_values) const
+    {
+        using ::std::get;
+        return __predicate(get<0>(__packed_neighbor_values), get<1>(__packed_neighbor_values));
+    }
+};
+
+template <typename _Predicate, typename _ValueType>
+struct __create_mask_unique_copy
+{
+    _Predicate __predicate;
+
+    template <typename _Idx, typename _Acc>
+    _ValueType
+    operator()(_Idx __idx, _Acc& __acc) const
+    {
+        using ::std::get;
+
+        auto __predicate_result = 1;
+        if (__idx != 0)
+            __predicate_result = __predicate(get<0>(__acc[__idx]), get<0>(__acc[__idx + (-1)]));
+
+        get<1>(__acc[__idx]) = __predicate_result;
+        return _ValueType{__predicate_result};
+    }
+};
+
+template <typename _ReduceValueType>
+struct __acc_handler_minelement
+{
+
+    template <typename _GlobalIdx, typename _Acc>
+    _ReduceValueType
+    operator()(_GlobalIdx __gidx, _Acc __acc) const
+    {
+        // TODO: __acc is tuple<accessors...> when zip_iterators are passed
+        return _ReduceValueType{__gidx, __acc[__gidx]};
+    }
+};
+
+template <typename _ReduceValueType>
+struct __acc_handler_minmaxelement
+{
+
+    template <typename _GlobalIdx, typename _Acc>
+    _ReduceValueType
+    operator()(_GlobalIdx __gidx, _Acc __acc) const
+    {
+        // TODO: Doesn't work with `zip_iterator`.
+        //       In that case the first and the second arguments of `_ReduceValueType` will be
+        //       a `tuple` of `difference_type`, not the `difference_type` itself.
+        return _ReduceValueType{__gidx, __gidx, __acc[__gidx], __acc[__gidx]};
+    }
+};
+
+template <typename _Compare>
+struct __identity_reduce_fn
+{
+    _Compare __comp;
+    template <typename _ReduceValueType>
+    _ReduceValueType
+    operator()(_ReduceValueType __a, _ReduceValueType __b) const
+    {
+        using ::std::get;
+        auto __chosen_for_min = __a;
+        auto __chosen_for_max = __b;
+        // if b "<" a or if b "==" a and b_index < a_index
+        if (__comp(get<2>(__b), get<2>(__a)) || (!__comp(get<2>(__a), get<2>(__b)) && get<0>(__b) < get<0>(__a)))
+            __chosen_for_min = __b;
+        // if a ">" b or if a "==" b and a_index > b_index
+        if (__comp(get<3>(__b), get<3>(__a)) || (!__comp(get<3>(__a), get<3>(__b)) && get<1>(__b) < get<1>(__a)))
+            __chosen_for_max = __a;
+        auto __result = _ReduceValueType{get<0>(__chosen_for_min), get<1>(__chosen_for_max), get<2>(__chosen_for_min),
+                                         get<3>(__chosen_for_max)};
+
+        return __result;
+    }
+};
+
+template <typename _Predicate>
+struct acc_handler_count
```

Comment:
Another candidate for uglification if needed.

Reply:
see my comment above